### PR TITLE
Update uart scheme

### PR DESCRIPTION
**Pull request by chasesza**

### CODE REVIEW

File: firmware/arduino/chessboard/uart.ino
Comment (nashirj) [inline firmware/arduino/chessboard/uart.ino:78]: Since you do this so often, consider making a function `isValidCoord(char c)` to do this check for you.
```
def isInvalidCoord(char c) {
    return c < 'A || c > 'L';
```

File: firmware/arduino/chessboard/uart.ino
Comment (nashirj) [inline firmware/arduino/chessboard/uart.ino:72]: - To simplify the amount of nested if statements/duplicated code for validation, you can validate the input at the beginning of the function. What you have is good, but just an alternate approach to separate validation from the other logic stuff
```
if (message[0] == ALIGN) {
    // validate start
    // set startPoint
}
if (message[0] == DIRECT || message[0] == EDGES) {
    // validate start and end
    // set startPoint, endPoint
}

if (message[0] == DIRECT) {
    // blah
} else if (message[0] == EDGES) {
    // blah
} else if (message[0] == ALIGN) {
    // blah
} else {
    // blah
}

// blah
```

File: firmware/arduino/chessboard/uart.ino
Comment (nashirj) [inline firmware/arduino/chessboard/uart.ino:63]: Should call `sendMessageToPi` to say that Arduino is executing before starting the movement function, since movement is blocking and it won't be able to tell the Pi that it's currently doing stuff once that whole thing starts. To that end, I think it could be cleaner if you are able to separate the parsing of the pi message from calling the movement function. I'll leave it up to you to decide if that's something you want to do, but this would allow getting the type of move/start and end points, then sending a message to the pi saying it's going to execute (or error if points are invalid), then calling the appropriate movement function. Some pseudocode:
```
if (parseMessageFromPi(buffer)) {
    sendMessageToPi(EXECUTING);
    // Here you would call appropriate move based on some global
    // variables you set in parseMessageFromPi
    doMove();  // blocks until move is finished, then sets status
} else {
    currentState = ERROR;
}
sendMessageToPi(currentState);  // can be either IDLE or ERROR
```

**Review by nashirj — COMMENTED** [review]
In addition to the other stuff, consider making a [`Point` variable to simplify parameter passing](https://github.com/ucfai/knightros-gambit/pull/70#pullrequestreview-839580635).

**Comment by nashirj** [comment]
PS, both you and @LSaber36 modified some of the same files, so let's wait to merge this in until his PR is merged since his came in first if that's ok.

**Comment by LSaber36** [comment]
Agreed

**Comment by LSaber36** [comment]
I'm getting in a call at 4:30 with John and Nashir, can you join us @chasesza

**Comment by chasesza** [comment]
> I'm getting in a call at 4:30 with John and Nashir, can you join us @chasesza

Yeah.

File: firmware/arduino/chessboard/uart.ino
Comment (nashirj) [inline firmware/arduino/chessboard/uart.ino:48]: This if statement will be true every time a new message comes in, since you set `currentState = IN_PROGRESS` right above this. Could maybe get rid of `IN_PROGRESS` state entirely, and just check `byte_num`
```c
if (byte_num != -1) {
   // blah
}
```

File: firmware/arduino/chessboard/uart.ino
Comment (nashirj) [inline firmware/arduino/chessboard/uart.ino:86]: This whole section can be simplified a bit:
```c
if (validateMessageFromPi(buffer)) {
    sendMessageToPi(EXECUTING, buffer[5], errorCode);
    // Note that makeMove() sets currentState to ERROR or IDLE accordingly
    makeMove(buffer);
}
sendMessageToPi(currentState, buffer[5], errorCode);
```

File: firmware/arduino/chessboard/uart.ino
Comment (nashirj) [inline firmware/arduino/chessboard/uart.ino:97]: Pls align the text in the second line with the first condition above, makes it easier to read.
```c
if (isInvalidCoord(message[1]) || isInvalidCoord(message[2]) ||
    isInvalidCoord(message[3]) || isInvalidCoord(message[4]))
```

File: firmware/arduino/chessboard/uart.ino
Comment (nashirj) [inline firmware/arduino/chessboard/uart.ino:133]: You can minimize code duplication here (if you want)
```c
// At start of function
bool moveSucceeded = false;
...
// For each movement case
moveSucceeded = moveX(...);
...
// After last else condition
if (!moveSucceeded) {
    currentState = ERROR;
    errorCode = MOVEMENT_ERROR;
    return false;
}
...
```
You don't have to do this, just a suggestion

**Review by nashirj — COMMENTED** [review]
I left a couple comments, but looks really good I'm ok with you merging after you address the comments and merge in @LSaber36 changes

File: firmware/arduino/chessboard/uart.ino
Comment (nashirj) [inline firmware/arduino/chessboard/uart.ino:10]: This indexing needs to be updated since you deleted `IN_PROGRESS`

**Review by nashirj — APPROVED** [review]
I'll merge as soon as you update the `ArduinoState` indexing

**Pull request merged**
